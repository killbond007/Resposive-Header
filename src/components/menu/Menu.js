import React, { useLayoutEffect, useRef, useState } from "react";

import PropTypes from "prop-types";

import * as Styled from "./__styles__/Menu.styles";

const Menu = ({ selectedKey, menuItems, onClick }) => {
  const menuRef = useRef(null);
  const menuItemsRef = useRef([]);

  const [visibleKeys, setVisibleKeys] = useState([]);

  useLayoutEffect(() => {
    handleResize();

    window.addEventListener("resize", handleResize, false);

    return () => {
      window.removeEventListener("resize", handleResize, false);
    };
  }, []);

  const handleResize = () => {
    let currentSumWidth, menuWidth;

    if (menuRef.current && menuRef.current.offsetWidth) {
      menuWidth = menuRef.current.offsetWidth;
      currentSumWidth =
        !!menuItems &&
        menuItems.reduce((acc, _, i) => {
          let width;
          const elem = menuItemsRef.current[i];

          if (elem) {
            const { marginLeft, marginRight } = getComputedStyle(elem);
            width =
              +marginLeft.replace("px", "") + +marginRight.replace("px", "");

            return acc + elem.offsetWidth + width;
          }
        }, 0);
      if (menuWidth < currentSumWidth) {
        let i = menuItems.length,
          keys = [];

        while (menuWidth < currentSumWidth && menuItemsRef.current[i - 1]) {
          currentSumWidth =
            currentSumWidth -
            menuItemsRef.current[i - 1].children[1].offsetWidth;

          keys.push((i - 1).toString());
          i--;
        }

        setVisibleKeys(keys);
      } else {
        let j = 0,
          keys = [];

        while (menuWidth > currentSumWidth + 120 && j < menuItems.length) {
          if (menuItemsRef.current[j].children[1].offsetWidth === 0) {
            keys.push(j.toString());
            currentSumWidth =
              currentSumWidth + menuItemsRef.current[j].children[1].offsetWidth;
            setVisibleKeys((prevState) =>
              prevState.filter((item) => item !== j.toString())
            );
          }
          j++;
        }
      }
    }
  };

  const _onMenuItemClick = (key) => () => onClick && onClick(key);

  return (
    <Styled.Root data-testid="__menu__" ref={menuRef}>
      {!!menuItems &&
        menuItems.map((item, i) => (
          <Styled.MenuItem
            data-testid="__menu-item__"
            key={item.key}
            $isSelected={selectedKey === item.key}
            ref={(ref) => (menuItemsRef.current[i] = ref)}
            onClick={_onMenuItemClick(item.key)}
          >
            {item.icon}
            <Styled.Title $isVisible={!visibleKeys.includes(i.toString())}>
              {item.title}
            </Styled.Title>
          </Styled.MenuItem>
        ))}
    </Styled.Root>
  );
};

Menu.propTypes = {
  /** key of currently selected menu items. */
  selectedKey: PropTypes.string,
  /** Array of MenuItems to appear in the menu. Each object is given as props to the MenuItem. */
  menuItems: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      title: PropTypes.string,
      icon: PropTypes.node,
    })
  ),
  /** Called when a menu item is clicked	. */
  onClick: PropTypes.func,
};

Menu.defaultProps = {
  selectedKey: null,
  menuItems: null,
  onClick: null,
};

export default Menu;
