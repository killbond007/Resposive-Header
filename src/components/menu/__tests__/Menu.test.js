import { fireEvent, render } from "@testing-library/react";
import React from "react";
import "jest-styled-components";

import Menu from "../Menu";
import DeleteIcon from "../../icon/DeleteIcon";

const getInstance = (props = {}, children = null) => (
  <Menu {...props}>{children}</Menu>
);

describe("Menu", () => {
  const menuItems = [
    { key: "navigation_1", title: "navigation_1", icon: <DeleteIcon /> },
    { key: "navigation_2", title: "navigation_2", icon: <DeleteIcon /> },
    { key: "navigation_3", title: "navigation_3", icon: <DeleteIcon /> },
    { key: "navigation_4", title: "navigation_4", icon: <DeleteIcon /> },
  ];

  describe("matches snapshot", () => {
    const data = [
      { title: "with no data" },
      { title: "with menuItems", props: { menuItems } },
    ];

    data.forEach(({ title, props }) => {
      it(title, () => {
        const { asFragment } = render(getInstance(props));
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  it("renders order menu items", () => {
    const { getAllByRole } = render(getInstance({ menuItems }));
    menuItems.map((item, i) => {
      expect(getAllByRole("listitem")[i]).toHaveTextContent(item.title);
    });
  });

  it("renders selected menu item", () => {
    const { getAllByTestId } = render(
      getInstance({
        selectedKey: "bar",
        menuItems: [
          { key: "foo", title: "Foo", icon: <DeleteIcon /> },
          { key: "bar", title: "Bar", icon: <DeleteIcon /> },
        ],
      })
    );
    expect(getAllByTestId("__menu-item__")[0]).toHaveStyleRule(
      "color",
      "#000000d9"
    );
    expect(getAllByTestId("__menu-item__")[1]).toHaveStyleRule(
      "color",
      "#1890ff"
    );
  });

  it("triggers onClick handler when clicking on menu item", () => {
    const onClick = jest.fn();
    const { getByText } = render(
      getInstance({
        onClick,
        menuItems: [
          { key: "foo", title: "Foo", icon: <DeleteIcon /> },
          { key: "bar", title: "Bar", icon: <DeleteIcon /> },
        ],
      })
    );
    fireEvent.click(getByText("Foo"));
    expect(onClick).toHaveBeenCalledWith("foo");
  });
});
