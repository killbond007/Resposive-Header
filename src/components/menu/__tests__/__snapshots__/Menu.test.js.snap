// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches snapshot with menuItems 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  line-height: 46px;
  height: 48px;
  border: 0;
  border-bottom: 1px solid #f0f0f0;
  box-shadow: none;
  text-align: left;
  list-style: none;
  overflow: hidden;
  margin: 0;
  padding: 0;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: -1px 20px 0;
  cursor: pointer;
  color: #000000d9;
}

.c2 {
  font-size: 14px;
  display: inline-block;
}

<ul
    class="c0"
    data-testid="__menu__"
  >
    <li
      class="c1"
      data-testid="__menu-item__"
    >
      <svg
        fill="currentColor"
        height="17px"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <span
        class="c2"
      >
        navigation_1
      </span>
    </li>
    <li
      class="c1"
      data-testid="__menu-item__"
    >
      <svg
        fill="currentColor"
        height="17px"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <span
        class="c2"
      >
        navigation_2
      </span>
    </li>
    <li
      class="c1"
      data-testid="__menu-item__"
    >
      <svg
        fill="currentColor"
        height="17px"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <span
        class="c2"
      >
        navigation_3
      </span>
    </li>
    <li
      class="c1"
      data-testid="__menu-item__"
    >
      <svg
        fill="currentColor"
        height="17px"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
      <span
        class="c2"
      >
        navigation_4
      </span>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Menu matches snapshot with no data 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  line-height: 46px;
  height: 48px;
  border: 0;
  border-bottom: 1px solid #f0f0f0;
  box-shadow: none;
  text-align: left;
  list-style: none;
  overflow: hidden;
  margin: 0;
  padding: 0;
}

<ul
    class="c0"
    data-testid="__menu__"
  />
</DocumentFragment>
`;
